File: scripts/deploy.py
"""
Deploy BitCred contracts to Starknet Sepolia testnet.

Usage:
    python deploy.py \
        --private-key 0x... \
        --account 0x...

Prerequisites:
    1. pip install starknet-py
    2. scarb build completed in ../contracts
    3. Account funded on Sepolia
"""

import asyncio
import argparse
import json
import os
from pathlib import Path
from starknet_py.net.full_node_client import FullNodeClient
from starknet_py.net.account.account import Account
from starknet_py.net.models.chains import StarknetChainId
from starknet_py.net.signer.stark_curve_signer import KeyPair
from starknet_py.contract import Contract

# ─── Config ───────────────────────────────────────────────────────────────────

RPC_ENDPOINTS = [
    os.getenv("STARKNET_RPC_URL"),
    "https://rpc.starknet-testnet.lava.build",
    "https://starknet-sepolia.public.blastapi.io/rpc/v0_9",
    "https://free-rpc.nethermind.io/sepolia-juno",
]

# Scarb outputs: contracts/target/dev/contracts_<Name>.contract_class.json
ARTIFACTS_DIR = Path(__file__).parent.parent / "contracts" / "target" / "dev"


# ─── Helpers ──────────────────────────────────────────────────────────────────

def load_contract(contract_name: str) -> tuple[dict, dict]:
    """Load Sierra + CASM from Scarb build artifacts."""
    sierra_path = ARTIFACTS_DIR / f"contracts_{contract_name}.contract_class.json"
    casm_path   = ARTIFACTS_DIR / f"contracts_{contract_name}.compiled_contract_class.json"

    if not sierra_path.exists():
        raise FileNotFoundError(f"Sierra not found: {sierra_path}\nRun 'scarb build' first.")
    if not casm_path.exists():
        raise FileNotFoundError(f"CASM not found: {casm_path}\nEnsure 'casm = true' in Scarb.toml")

    with open(sierra_path) as f:
        sierra = json.load(f)
    with open(casm_path) as f:
        casm = json.load(f)

    return sierra, casm


async def get_working_client() -> FullNodeClient:
    """Try RPC endpoints in order, return first that responds."""
    for url in filter(None, RPC_ENDPOINTS):
        try:
            client = FullNodeClient(node_url=url)
            await client.get_block_number()
            print(f"   ✅ Connected: {url}\n")
            return client
        except Exception:
            print(f"   ❌ Unreachable: {url}")
    raise ConnectionError("All RPC endpoints failed.")


# ─── Declare + Deploy ─────────────────────────────────────────────────────────

async def declare_and_deploy(account: Account, contract_name: str, constructor_args: list) -> int:
    """Declare if needed, then deploy. Returns deployed contract address."""
    print(f"📝 Declaring {contract_name}...")
    sierra, casm = load_contract(contract_name)

    declare_result = await Contract.declare_v3(
        account=account,
        compiled_contract=json.dumps(sierra),
        compiled_contract_casm=json.dumps(casm),
    )
    await declare_result.wait_for_acceptance()
    print(f"   Class hash: {hex(declare_result.class_hash)}")

    print(f"🚀 Deploying {contract_name}...")
    print(f"   Constructor args: {[hex(a) if isinstance(a, int) else a for a in constructor_args]}")

    deploy_result = await declare_result.deploy_v3(constructor_args=constructor_args)
    await deploy_result.wait_for_acceptance()

    address = deploy_result.deployed_contract.address
    print(f"   ✅ Address: {hex(address)}\n")
    return address


# ─── Main ─────────────────────────────────────────────────────────────────────

async def deploy(private_key: str, account_address: str):
    print("\n🏗️  BITCRED PYTHON DEPLOY SCRIPT")
    print("=" * 50)
    print(f"Account: {account_address}")
    print(f"Network: Starknet Sepolia\n")

    print("🔌 Finding working RPC endpoint...")
    client = await get_working_client()

    account = Account(
        client=client,
        address=int(account_address, 16),
        key_pair=KeyPair.from_private_key(int(private_key, 16)),
        chain=StarknetChainId.SEPOLIA,
    )

    # ── 1. ScoreRegistry ──────────────────────────────────────────────────────
    # constructor(admin: ContractAddress)
    registry_address = await declare_and_deploy(
        account=account,
        contract_name="ScoreRegistry",
        constructor_args=[int(account_address, 16)],  # admin = deployer
    )

    # ── 2. LendingPool ────────────────────────────────────────────────────────
    # constructor(admin, score_registry, collateral_token, borrow_token, interest_rate)
    wbtc_address  = int(os.getenv("WBTC_ADDRESS",  "0x0"), 16)
    usdc_address  = int(os.getenv("USDC_ADDRESS",  "0x0"), 16)
    interest_rate = 500  # 5% annual in basis points

    lending_address = await declare_and_deploy(
        account=account,
        contract_name="LendingPool",
        constructor_args=[
            int(account_address, 16),  # admin
            registry_address,          # score_registry
            wbtc_address,              # collateral_token
            usdc_address,              # borrow_token
            interest_rate,             # initial_interest_rate
        ],
    )

    # ── 3. Save to backend/.env ───────────────────────────────────────────────
    env_path = Path(__file__).parent.parent / "backend" / ".env"
    env_path.parent.mkdir(parents=True, exist_ok=True)

    with open(env_path, "w") as f:
        f.write(f"STARKNET_RPC_URL={client.url}\n")
        f.write(f"REGISTRY_ADDRESS={hex(registry_address)}\n")
        f.write(f"LENDING_ADDRESS={hex(lending_address)}\n")
        f.write(f"SCORER_ACCOUNT_ADDR={account_address}\n")
        f.write(f"SCORER_PRIVATE_KEY={private_key}\n")

    print("=" * 50)
    print("ALL CONTRACTS DEPLOYED 🎉")
    print("=" * 50)
    print(f"\n  SCORE_REGISTRY: {hex(registry_address)}")
    print(f"  LENDING_POOL:   {hex(lending_address)}")
    print(f"\n💾 Saved to {env_path}")
    print("\n📋 Next steps:")
    print("  1. Add WBTC_ADDRESS + USDC_ADDRESS to backend/.env")
    print("  2. pip install -r backend/requirements.txt")
    print("  3. uvicorn backend.api.main:app --reload")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Deploy BitCred contracts to Starknet Sepolia")
    parser.add_argument("--private-key", required=True, help="Starknet account private key (0x...)")
    parser.add_argument("--account",     required=True, help="Starknet account address (0x...)")
    args = parser.parse_args()

    asyncio.run(deploy(args.private_key, args.account))

File: scripts/deploy.js
const {
  RpcProvider,
  Account,
  ETransactionVersion,
  defaultDeployer,
  json,
  hash,
} = require("starknet");
const fs = require("fs");
require("dotenv").config();

const account_address = process.env.ACCOUNT_ADDRESS;
const private_key = process.env.PRIVATE_KEY;

if (!account_address || !private_key) {
  console.error("❌ Missing ACCOUNT_ADDRESS or PRIVATE_KEY in .env");
  process.exit(1);
}

const RPC_ENDPOINTS = [
  process.env.STARKNET_RPC_URL,
  "https://starknet-sepolia.g.alchemy.com/v2/demo",
  "https://starknet-sepolia.infura.io/v3/9aa3d95b3bc440fa88ea12eaa4456161",
  "https://free-rpc.nethermind.io/sepolia-juno",
  "https://rpc.starknet-testnet.lava.build",
].filter(Boolean);

async function getWorkingProvider() {
  for (const url of RPC_ENDPOINTS) {
    try {
      console.log(`   Trying: ${url}`);
      const p = new RpcProvider({ nodeUrl: url, retries: 2 });
      
      // Try a simpler method first
      try {
        await p.getChainId();
        console.log(`   ✅ Connected (chain ID: ${await p.getChainId()}!\n`);
        return p;
      } catch {
        // Fallback to block number check
        await p.getBlockNumber();
        console.log(`   ✅ Connected!\n`);
        return p;
      }
    } catch {
      console.log(`   ❌ Unreachable\n`);
    }
  }
  throw new Error("All RPC endpoints failed. Check your internet connection.");
}

const CONTRACTS = {
  SCORE_REGISTRY: "ScoreRegistry",
  LENDING_POOL:   "LendingPool",
  MOCK_WBTC:      "MockWBTC",      
  MOCK_USDC:      "MockUSDC",      
};

const CLASS_HASHES = {
  SCORE_REGISTRY: "",
  LENDING_POOL:   "",
  MOCK_WBTC:      "",              
  MOCK_USDC:      "",              
};

const DEPLOYED_ADDRESSES = {
  SCORE_REGISTRY: "",
  LENDING_POOL:   "",
  MOCK_WBTC:      "",              
  MOCK_USDC:      "",              
};

function loadContract(contractName) {
  const sierraPath = `../contracts/target/dev/contracts_${contractName}.contract_class.json`;
  const casmPath   = `../contracts/target/dev/contracts_${contractName}.compiled_contract_class.json`;

  if (!fs.existsSync(sierraPath)) {
    throw new Error(`Sierra file not found: ${sierraPath}\nRun 'scarb build' first.`);
  }
  if (!fs.existsSync(casmPath)) {
    throw new Error(`CASM file not found: ${casmPath}\nEnsure 'casm = true' is in Scarb.toml`);
  }

  const compiledSierra = json.parse(fs.readFileSync(sierraPath).toString("ascii"));
  const compiledCasm   = json.parse(fs.readFileSync(casmPath).toString("ascii"));
  return { compiledSierra, compiledCasm };
}

async function declareContract(provider, account, contractName) {
  console.log(`\n📝 Declaring ${contractName}...`);

  const { compiledSierra, compiledCasm } = loadContract(contractName);

  const classHash = hash.computeContractClassHash(compiledSierra);
  console.log(`   Computed class hash: ${classHash}`);

  try {
    await provider.getClassByHash(classHash);
    console.log(`   ✅ Already declared — skipping.`);
    return classHash;
  } catch {
    console.log(`   🔄 Not yet declared, submitting...`);
  }

  const declareResponse = await account.declare({
    contract: compiledSierra,
    casm: compiledCasm,
  });

  console.log(`   Tx hash: ${declareResponse.transaction_hash}`);
  await provider.waitForTransaction(declareResponse.transaction_hash);

  console.log(`   ✅ Declared! Class hash: ${declareResponse.class_hash}`);
  return declareResponse.class_hash;
}

async function deployContract(provider, account, contractName, classHash, constructorArgs = []) {
  console.log(`\n🚀 Deploying ${contractName}...`);
  console.log(`   Class hash: ${classHash}`);
  console.log(`   Constructor args:`, constructorArgs);

  await new Promise(r => setTimeout(r, 2000));

  const deployResponse = await account.deployContract({
    classHash,
    constructorCalldata: constructorArgs,
  });

  console.log(`   Tx hash: ${deployResponse.transaction_hash}`);

  await provider.waitForTransaction(deployResponse.transaction_hash, {
    retryInterval: 5000,
    successStates: ["ACCEPTED_ON_L2", "ACCEPTED_ON_L1"],
  });

  const contractAddress = Array.isArray(deployResponse.contract_address)
    ? deployResponse.contract_address[0]
    : deployResponse.contract_address;

  if (!contractAddress || contractAddress === "0x0") {
    throw new Error(`Deployment returned invalid address for ${contractName}`);
  }

  console.log(`   ✅ Deployed! Address: ${contractAddress}`);
  return contractAddress;
}

async function declareAll(provider, account) {
  console.log("\n" + "=".repeat(60));
  console.log("STEP 1: DECLARING CONTRACTS");
  console.log("=".repeat(60));

  CLASS_HASHES.SCORE_REGISTRY = await declareContract(provider, account, CONTRACTS.SCORE_REGISTRY);
  await new Promise(r => setTimeout(r, 5000));

  CLASS_HASHES.MOCK_WBTC = await declareContract(provider, account, CONTRACTS.MOCK_WBTC);
  await new Promise(r => setTimeout(r, 5000));

  CLASS_HASHES.MOCK_USDC = await declareContract(provider, account, CONTRACTS.MOCK_USDC);
  await new Promise(r => setTimeout(r, 5000));

  CLASS_HASHES.LENDING_POOL = await declareContract(provider, account, CONTRACTS.LENDING_POOL);

  console.log("\n" + "=".repeat(60));
  console.log("ALL CONTRACTS DECLARED ✅");
  console.log("=".repeat(60));
  console.log("\nClass Hashes:");
  console.log("  SCORE_REGISTRY:", CLASS_HASHES.SCORE_REGISTRY);
  console.log("  MOCK_WBTC:     ", CLASS_HASHES.MOCK_WBTC);
  console.log("  MOCK_USDC:     ", CLASS_HASHES.MOCK_USDC);
  console.log("  LENDING_POOL:  ", CLASS_HASHES.LENDING_POOL);
}


async function deployAll(provider, account) {
  console.log("\n" + "=".repeat(60));
  console.log("STEP 2: DEPLOYING CONTRACTS");
  console.log("=".repeat(60));

  // Deploy ScoreRegistry
  DEPLOYED_ADDRESSES.SCORE_REGISTRY = await deployContract(
    provider, account,
    CONTRACTS.SCORE_REGISTRY,
    CLASS_HASHES.SCORE_REGISTRY,
    [account_address],
  );
  await new Promise(r => setTimeout(r, 10000));

  // Deploy Mock Tokens
  DEPLOYED_ADDRESSES.MOCK_WBTC = await deployContract(
    provider, account,
    CONTRACTS.MOCK_WBTC,
    CLASS_HASHES.MOCK_WBTC,
    [],
  );
  await new Promise(r => setTimeout(r, 10000));

  DEPLOYED_ADDRESSES.MOCK_USDC = await deployContract(
    provider, account,
    CONTRACTS.MOCK_USDC,
    CLASS_HASHES.MOCK_USDC,
    [],
  );
  await new Promise(r => setTimeout(r, 10000));

  // Deploy LendingPool with mock token addresses
  DEPLOYED_ADDRESSES.LENDING_POOL = await deployContract(
    provider, account,
    CONTRACTS.LENDING_POOL,
    CLASS_HASHES.LENDING_POOL,
    [
      account_address,
      DEPLOYED_ADDRESSES.SCORE_REGISTRY,
      DEPLOYED_ADDRESSES.MOCK_WBTC,    // Use our mock WBTC
      DEPLOYED_ADDRESSES.MOCK_USDC,    // Use our mock USDC
      500,                              // 5% interest
    ],
  );

  const deploymentInfo = {
    network:     "starknet-sepolia",
    timestamp:   new Date().toISOString(),
    deployer:    account_address,
    classHashes: CLASS_HASHES,
    addresses:   DEPLOYED_ADDRESSES,
  };

  fs.writeFileSync("./deployment-info.json", JSON.stringify(deploymentInfo, null, 2));

  console.log("\n" + "=".repeat(60));
  console.log("ALL CONTRACTS DEPLOYED 🎉");
  console.log("=".repeat(60));
  console.log("\nDeployed Addresses:");
  console.log("  SCORE_REGISTRY:", DEPLOYED_ADDRESSES.SCORE_REGISTRY);
  console.log("  MOCK_WBTC:     ", DEPLOYED_ADDRESSES.MOCK_WBTC);
  console.log("  MOCK_USDC:     ", DEPLOYED_ADDRESSES.MOCK_USDC);
  console.log("  LENDING_POOL:  ", DEPLOYED_ADDRESSES.LENDING_POOL);
  console.log("\n💾 Saved to deployment-info.json");
  console.log("\n📋 Update your .env and frontend:");
  console.log(`  REGISTRY_ADDRESS=${DEPLOYED_ADDRESSES.SCORE_REGISTRY}`);
  console.log(`  LENDING_ADDRESS=${DEPLOYED_ADDRESSES.LENDING_POOL}`);
  console.log(`  MOCK_WBTC_ADDRESS=${DEPLOYED_ADDRESSES.MOCK_WBTC}`);
  console.log(`  MOCK_USDC_ADDRESS=${DEPLOYED_ADDRESSES.MOCK_USDC}`);
}

async function verifyDeployments(provider) {
  console.log("\n🔍 Verifying deployments on-chain...\n");

  let info;
  try {
    info = JSON.parse(fs.readFileSync("./deployment-info.json").toString());
  } catch {
    console.log("⚠️  deployment-info.json not found — skipping verification.");
    return;
  }

  for (const [name, address] of Object.entries(info.addresses)) {
    try {
      const classHash = await provider.getClassHashAt(address);
      console.log(`✅ ${name}`);
      console.log(`   Address:    ${address}`);
      console.log(`   Class hash: ${classHash}`);
    } catch {
      console.log(`❌ ${name}: ${address} — NOT FOUND on chain`);
    }
  }
}

async function checkBalance(provider, address) {
  try {
    // Try different methods to get balance
    try {
      const balance = await provider.getBalance(address);
      console.log("Balance:", balance.toString());
      return;
    } catch {
      // Fallback: try to get the account's nonce as a sign it's alive
      const nonce = await provider.getNonce(address);
      console.log("Account nonce:", nonce);
    }
  } catch {
    console.log("⚠️  Could not fetch account info");
  }
}

async function main() {
  console.log("\n🏗️  BITCRED DEPLOYMENT SCRIPT");
  console.log("================================");
  console.log("Account:", account_address);
  console.log("Network: Starknet Sepolia\n");

  console.log("🔌 Finding working RPC endpoint...");
  let provider;
  try {
    provider = await getWorkingProvider();
  } catch (err) {
    console.error(`\n❌ ${err.message}`);
    process.exit(1);
  }

  const account = new Account({
    provider,
    address: account_address,
    signer: private_key,
    cairoVersion: "1",
    transactionVersion: ETransactionVersion.V3,
    deployer: defaultDeployer,
  });

  // Check account info
  console.log("\n📊 Account Info:");
  await checkBalance(provider, account_address);
  console.log("");

  // Step 1: Declare
  await declareAll(provider, account);

  // Step 2: Deploy
  await deployAll(provider, account);

  // Step 3: Verify
  await verifyDeployments(provider);
}

main().catch(err => {
  console.error("\n❌ Deployment failed:", err.message);
  console.error(err.stack);
  process.exit(1);
});